File: src/ruby/tormenta20/seeder.rb
# frozen_string_literal: true

require "json"

module Tormenta20
  # Handles importing JSON data into the database
  module Seeder
    JSON_BASE_PATH = File.expand_path("../../json", __dir__)

    class << self
      def seed_all(verbose: false)
        @verbose = verbose
        log "Seeding database..."

        seed_racas
        seed_habilidades_de_raca
        seed_origens
        seed_poderes_habilidades_unicas
        seed_divindades
        seed_poderes_concedidos
        seed_poderes_tormenta
        seed_poderes_gerais
        seed_classes
        seed_habilidades_de_classe
        seed_magias
        seed_equipamentos
        seed_itens_superiores
        seed_regras
        seed_condicoes

        log "Database seeded successfully!"
      end

      def seeded?
        Models::Origem.count.positive?
      rescue StandardError
        false
      end

      private

      def log(message)
        puts message if @verbose
      end

      def seed_racas
        import_json_files(
          "racas",
          Models::Raca,
          %i[id name description size movement vision vision_range
             attribute_bonuses skill_bonuses racial_abilities
             chosen_abilities_amount available_chosen_abilities]
        )
      end

      def seed_habilidades_de_raca
        base_dir = File.join(JSON_BASE_PATH, "poderes", "habilidades_de_raca")
        return unless Dir.exist?(base_dir)

        Dir.each_child(base_dir) do |race_dir|
          race_path = File.join(base_dir, race_dir)
          next unless File.directory?(race_path)

          Dir.glob(File.join(race_path, "*.json")).each do |file|
            data = JSON.parse(File.read(file), symbolize_names: true)
            next if data[:id].nil?

            record = Models::Poder.find_or_initialize_by(id: data[:id])
            record.name        = data[:name] || data[:id].to_s.humanize
            record.type        = "habilidade_de_raca"
            record.description = data[:description]
            record.effects     = data[:effects] || {}
            record.prerequisites = data[:prerequisites] || []
            record.save!
          rescue StandardError
            next
          end
        end
      end

      def seed_origens
        import_json_files(
          "origens",
          Models::Origem,
          %i[id name description items benefits unique_power]
        )
      end

      def seed_poderes_habilidades_unicas
        import_json_files(
          "poderes/habilidades_unicas_de_origem",
          Models::Poder,
          %i[id name description effects prerequisites],
          extra_attrs: { type: "habilidade_unica_origem" },
          transform: lambda { |data|
            data[:origin_id] = data.delete(:origin) if data[:origin]
            data
          }
        )
      end

      def seed_divindades
        import_json_files(
          "deuses",
          Models::Divindade,
          %i[id name title description beliefs_objectives holy_symbol energy
             preferred_weapon devotees granted_powers obligations_restrictions]
        )
      end

      def seed_poderes_concedidos
        import_json_files(
          "poderes/poderes_concedidos",
          Models::Poder,
          %i[id name description effects prerequisites deities],
          extra_attrs: { type: "poder_concedido" }
        )
      end

      def seed_poderes_tormenta
        import_json_files(
          "poderes/poderes_da_tormenta",
          Models::Poder,
          %i[id name description effects prerequisites],
          extra_attrs: { type: "poder_tormenta" }
        )
      end

      def seed_poderes_gerais
        base_dir = File.join(JSON_BASE_PATH, "poderes", "poderes_gerais")
        return unless Dir.exist?(base_dir)

        Dir.glob(File.join(base_dir, "**", "*.json")).each do |file|
          data = JSON.parse(File.read(file), symbolize_names: true)
          next if data[:id].nil?

          record = Models::Poder.find_or_initialize_by(id: data[:id])
          record.name        = data[:name] || data[:id].to_s.humanize
          record.type        = data[:type] || "poder_geral"
          record.description = data[:description]
          record.effects     = data[:effects] || []
          record.costs       = data[:costs] || []
          record.prerequisites = data[:requirements] || []
          record.save!
        rescue StandardError
          next
        end
      end

      def seed_classes
        import_json_files(
          "classes",
          Models::Classe,
          %i[id name hit_points mana_points skills proficiencies abilities powers progression spellcasting]
        )
      end

      def seed_habilidades_de_classe
        base_dir = File.join(JSON_BASE_PATH, "poderes", "habilidades_de_classe")
        return unless Dir.exist?(base_dir)

        Dir.each_child(base_dir) do |class_dir|
          class_path = File.join(base_dir, class_dir)
          next unless File.directory?(class_path)
          next if class_dir == "base.json"

          class_id = class_dir

          Dir.glob(File.join(class_path, "*.json")).each do |file|
            data = JSON.parse(File.read(file), symbolize_names: true)
            next if data[:id].nil?

            record = Models::Poder.find_or_initialize_by(id: data[:id])
            record.name        = data[:name] || data[:id].to_s.humanize
            record.type        = "poder_classe"
            record.class_id    = class_id
            record.description = data[:description]
            record.effects     = data[:effects] || []
            record.costs       = data[:costs] || []
            record.prerequisites = data[:requirements] || []
            record.save!
          rescue StandardError
            next
          end
        end
      end

      def seed_magias
        json_dir = File.join(JSON_BASE_PATH, "magias")
        return unless Dir.exist?(json_dir)

        Dir.glob(File.join(json_dir, "*.json")).each do |file|
          data = JSON.parse(File.read(file), symbolize_names: true)
          import_magia(data)
        rescue StandardError
          next
        end
      end

      def import_magia(data)
        raw_target = data[:target]
        target = if raw_target.is_a?(Hash)
                   raw_target
                 else
                   (raw_target.is_a?(Array) ? raw_target.first || {} : {})
                 end
        effect_details = data[:effect_details].is_a?(Hash) ? data[:effect_details] : {}
        resistence = data[:resistence].is_a?(Hash) ? data[:resistence] : {}
        extra_costs = data[:extra_costs].is_a?(Hash) ? data[:extra_costs] : {}

        Models::Magia.find_or_initialize_by(id: data[:id]).tap do |magia|
          magia.name = data[:name]
          magia.type = data[:type]
          magia.circle = data[:circle]
          magia.school = data[:school]
          magia.execution = data[:execution]
          magia.execution_details = data[:execution_details]
          magia.range = data[:range]
          magia.duration = data[:duration]
          magia.duration_details = data[:duration_details]
          magia.counterspell = data[:counterspell]
          magia.description = data[:description] || ""

          magia.target_amount = target[:amount]
          magia.target_up_to = if target[:up_to].nil?
                                 nil
                               else
                                 (target[:up_to] ? 1 : 0)
                               end
          magia.target_type = target[:type]

          magia.effect = data[:effect]
          magia.effect_shape = effect_details[:shape]
          magia.effect_dimention = effect_details[:dimention]
          magia.effect_size = effect_details[:size]
          magia.effect_other_details = effect_details[:other_details]

          magia.area_effect = data[:area_effect]
          magia.area_effect_details = data[:area_effect_details]

          magia.resistence_effect = resistence[:effect]
          magia.resistence_skill = resistence[:skill]

          magia.extra_costs_material_component = extra_costs[:material_component]
          magia.extra_costs_material_cost = extra_costs[:material_component_cost]
          magia.extra_costs_pm_debuff = extra_costs[:pm_debuff]
          magia.extra_costs_pm_sacrifice = extra_costs[:pm_sacrifice]

          magia.enhancements = data[:enhancements] || []
          magia.effects = data[:effects] || []

          magia.save!
        end
      end

      def seed_equipamentos
        seed_armas
        seed_armaduras
        seed_escudos
        seed_itens
      end

      def seed_armas
        import_json_files(
          "equipamentos/armas",
          Models::Arma,
          %i[id name category price damage damage_type critical range weight properties description]
        )
      end

      def seed_armaduras
        import_json_files(
          "equipamentos/armaduras",
          Models::Armadura,
          %i[id name category price defense_bonus armor_penalty weight properties description]
        )
      end

      def seed_escudos
        import_json_files(
          "equipamentos/escudos",
          Models::Escudo,
          %i[id name price defense_bonus armor_penalty weight properties description],
          transform: lambda { |data|
            data[:price] ||= data.delete(:preco)
            data[:defense_bonus] ||= data.delete(:bonus_defesa)
            data[:armor_penalty] ||= data.delete(:penalidade_armadura)
            data[:weight] ||= data.delete(:espacos)
            data
          }
        )
      end

      def seed_itens
        import_json_files(
          "equipamentos/itens",
          Models::Item,
          %i[id name category price weight description effects]
        )
      end

      def seed_itens_superiores
        seed_materiais_especiais
        seed_melhorias
      end

      def seed_materiais_especiais
        import_json_files(
          "itens_superiores/materiais_especiais",
          Models::MaterialEspecial,
          %i[id name description applicable_to price_modifier effects]
        )
      end

      def seed_melhorias
        import_json_files(
          "itens_superiores/melhorias",
          Models::Melhoria,
          %i[id name description applicable_to price effects]
        )
      end

      def seed_regras
        import_json_files(
          "regras",
          Models::Regra,
          %i[id name description data],
          transform: lambda { |raw|
            base_keys = %i[id name description]
            data_content = raw.reject { |k, _| base_keys.include?(k) }
            { id: raw[:id], name: raw[:name], description: raw[:description], data: data_content }
          }
        )
      end

      def seed_condicoes
        import_json_files(
          "condicoes",
          Models::Condicao,
          %i[id name description effects condition_type escalates_to]
        )
      end

      def import_json_files(subdir, model, attrs, extra_attrs: {}, transform: nil)
        json_dir = File.join(JSON_BASE_PATH, subdir)
        return unless Dir.exist?(json_dir)

        Dir.glob(File.join(json_dir, "**", "*.json")).each do |file|
          data = JSON.parse(File.read(file), symbolize_names: true)
          data = transform.call(data) if transform

          record = model.find_or_initialize_by(id: data[:id])
          attrs.each do |attr|
            record.send("#{attr}=", data[attr]) if data.key?(attr)
          end
          extra_attrs.each do |attr, value|
            record.send("#{attr}=", value)
          end
          record.save!
        rescue StandardError
          next
        end
      end
    end
  end
end

File: src/ts/seeder.ts
import Database from "better-sqlite3"
import { readdirSync, readFileSync, statSync } from "fs"
import { join, dirname, basename } from "path"
import { fileURLToPath } from "url"

const __dirname = dirname(fileURLToPath(import.meta.url))
const JSON_DIR = join(__dirname, "../../src/json")
const DB_PATH = process.env.TORMENTA20_DB_PATH ?? join(__dirname, "../../db/tormenta20.sqlite3")
const SCHEMA_PATH = join(__dirname, "../../db/schema.sql")

function readJson<T>(path: string): T {
  return JSON.parse(readFileSync(path, "utf-8")) as T
}

function loadJsonFiles<T>(dir: string): T[] {
  const results: T[] = []
  for (const entry of readdirSync(dir)) {
    const full = join(dir, entry)
    if (statSync(full).isDirectory()) {
      results.push(...loadJsonFiles<T>(full))
    } else if (entry.endsWith(".json")) {
      const parsed = readJson<T>(full)
      if (parsed && typeof parsed === "object" && "id" in (parsed as object) && (parsed as Record<string, unknown>).id !== "") {
        results.push(parsed)
      }
    }
  }
  return results
}

function loadJsonDir<T>(dir: string): T[] {
  return readdirSync(dir)
    .filter((f) => f.endsWith(".json"))
    .map((f) => readJson<T>(join(dir, f)))
    .filter((p) => (p as Record<string, unknown>).id !== "")
}

export function seed(): void {
  const db = new Database(DB_PATH)
  db.pragma("journal_mode = WAL")
  db.pragma("foreign_keys = OFF")

  const schema = readFileSync(SCHEMA_PATH, "utf-8")
  db.exec(schema)

  const insert = db.transaction((table: string, rows: Record<string, unknown>[]) => {
    if (rows.length === 0) return
    const keys = Object.keys(rows[0])
    const cols = keys.join(", ")
    const vals = keys.map(() => "?").join(", ")
    const stmt = db.prepare(`INSERT OR REPLACE INTO ${table} (${cols}) VALUES (${vals})`)
    for (const row of rows) stmt.run(...keys.map((k) => {
      const v = row[k]
      return typeof v === "object" && v !== null ? JSON.stringify(v) : v
    }))
  })

  seedRacas(insert)
  seedOrigens(insert)
  seedDivindades(insert)
  seedClasses(insert)
  seedPoderes(insert)
  seedMagias(insert)
  seedEquipamentos(insert)
  seedItensSuperiores(insert)
  seedCondicoes(insert)
  seedPericias(insert)
  seedRegras(insert)

  db.pragma("foreign_keys = ON")
  db.close()
  console.log("✓ Tormenta 20 database seeded successfully")
}

function seedRacas(insert: ReturnType<typeof Database.prototype.transaction>): void {
  const racas = loadJsonDir<Record<string, unknown>>(join(JSON_DIR, "racas"))
  insert("racas", racas)
}

function seedOrigens(insert: ReturnType<typeof Database.prototype.transaction>): void {
  const origens = loadJsonDir<Record<string, unknown>>(join(JSON_DIR, "origens"))
  insert("origens", origens.map((o) => ({
    id: o.id, name: o.name, description: o.description ?? null,
    items: o.items ?? [], benefits: o.benefits ?? {}, unique_power: o.unique_power ?? null,
  })))
}

function seedDivindades(insert: ReturnType<typeof Database.prototype.transaction>): void {
  const divs = loadJsonDir<Record<string, unknown>>(join(JSON_DIR, "deuses"))
  insert("divindades", divs.map((d) => ({
    id: d.id, name: d.name, title: d.title ?? null, description: d.description ?? null,
    beliefs_objectives: d.beliefs_objectives ?? [], holy_symbol: d.holy_symbol ?? null,
    energy: d.energy ?? "qualquer", preferred_weapon: d.preferred_weapon ?? null,
    devotees: d.devotees ?? {}, granted_powers: d.granted_powers ?? [],
    obligations_restrictions: d.obligations_restrictions ?? null,
  })))
}

function seedClasses(insert: ReturnType<typeof Database.prototype.transaction>): void {
  const classes = loadJsonDir<Record<string, unknown>>(join(JSON_DIR, "classes"))
  insert("classes", classes.map((c) => ({
    id: c.id, name: c.name,
    hit_points: c.hit_points ?? {}, mana_points: c.mana_points ?? {},
    skills: c.skills ?? {}, proficiencies: c.proficiencies ?? {},
    abilities: c.abilities ?? [], powers: c.powers ?? [],
    progression: c.progression ?? [],
    spellcasting: c.spellcasting && Object.keys(c.spellcasting as object).length > 0 ? c.spellcasting : null,
  })))
}

const PODER_DIR_TYPE_MAP: Record<string, string> = {
  habilidades_de_raca: "habilidade_de_raca",
  habilidades_unicas_de_origem: "habilidade_unica_origem",
  poderes_concedidos: "poder_concedido",
  poderes_da_tormenta: "poder_tormenta",
  poderes_de_combate: "poder_combate",
  poderes_de_destino: "poder_destino",
  poderes_de_magia: "poder_magia",
}

function seedPoderes(insert: ReturnType<typeof Database.prototype.transaction>): void {
  const poderesDir = join(JSON_DIR, "poderes")
  const rows: Record<string, unknown>[] = []

  for (const topDir of readdirSync(poderesDir)) {
    const topPath = join(poderesDir, topDir)
    if (!statSync(topPath).isDirectory()) continue

    if (topDir === "poderes_gerais") {
      for (const subDir of readdirSync(topPath)) {
        const subPath = join(topPath, subDir)
        if (!statSync(subPath).isDirectory()) continue
        const type = PODER_DIR_TYPE_MAP[subDir] ?? "poder_geral"
        const files = loadJsonFiles<Record<string, unknown>>(subPath)
        rows.push(...files.map((p) => mapPoder(p, type, null, null)))
      }
      continue
    }

    if (topDir === "habilidades_de_classe") {
      for (const classDir of readdirSync(topPath)) {
        const classPath = join(topPath, classDir)
        if (!statSync(classPath).isDirectory()) continue
        const files = loadJsonDir<Record<string, unknown>>(classPath)
        rows.push(...files.map((p) => mapPoder(p, "poder_classe", null, classDir)))
      }
      continue
    }

    const type = PODER_DIR_TYPE_MAP[topDir] ?? "poder_geral"

    if (topDir === "habilidades_de_raca") {
      for (const subDir of readdirSync(topPath)) {
        const subPath = join(topPath, subDir)
        if (statSync(subPath).isDirectory()) {
          const files = loadJsonDir<Record<string, unknown>>(subPath)
          rows.push(...files.map((p) => mapPoder(p, type, null, null)))
        } else if (subDir.endsWith(".json")) {
          const p = readJson<Record<string, unknown>>(subPath)
          if ((p.id as string) !== "") rows.push(mapPoder(p, type, null, null))
        }
      }
      continue
    }

    if (topDir === "habilidades_unicas_de_origem") {
      const files = loadJsonDir<Record<string, unknown>>(topPath)
      rows.push(...files.map((p) => mapPoder(p, type, (p.origin as string) ?? null, null)))
      continue
    }

    const files = loadJsonFiles<Record<string, unknown>>(topPath)
    rows.push(...files.map((p) => mapPoder(p, type, null, null)))
  }

  insert("poderes", rows)
}

function mapPoder(p: Record<string, unknown>, type: string, originId: string | null, classId: string | null): Record<string, unknown> {
  return {
    id: p.id, name: p.name, type,
    description: p.description ?? null,
    effects: p.effects ?? [], costs: p.costs ?? [],
    prerequisites: p.requirements ?? p.prerequisites ?? [],
    origin_id: originId, class_id: classId,
    deities: p.deities ?? [],
  }
}

function seedMagias(insert: ReturnType<typeof Database.prototype.transaction>): void {
  const magias = loadJsonDir<Record<string, unknown>>(join(JSON_DIR, "magias"))
  insert("magias", magias.map((m) => {
    const target = Array.isArray(m.target) ? m.target[0] : (m.target as Record<string, unknown> | null)
    return {
      id: m.id, name: m.name, type: m.type, circle: m.circle, school: m.school,
      execution: m.execution, execution_details: m.execution_details ?? null,
      range: m.range, duration: m.duration, duration_details: m.duration_details ?? null,
      counterspell: m.counterspell ?? null, description: m.description,
      target_amount: target?.amount ?? null,
      target_up_to: target?.up_to ? 1 : 0,
      target_type: target?.type ?? null,
      effect: m.effect ?? null, effect_shape: null, effect_dimention: null,
      effect_size: null, effect_other_details: m.effect_details ?? null,
      area_effect: null, area_effect_details: null,
      resistence_effect: (m.resistence as Record<string, unknown>)?.effect ?? null,
      resistence_skill: (m.resistence as Record<string, unknown>)?.skill ?? null,
      extra_costs_material_component: null, extra_costs_material_cost: null,
      extra_costs_pm_debuff: null, extra_costs_pm_sacrifice: null,
      enhancements: m.enhancements ?? [], effects: m.effects ?? [],
    }
  }))
}

const ARMA_CATEGORY_MAP: Record<string, string> = {
  simples: "simples",
  marcial: "marciais",
  exotica: "exoticas",
  fogo: "fogo",
}

function seedEquipamentos(insert: ReturnType<typeof Database.prototype.transaction>): void {
  const eqDir = join(JSON_DIR, "equipamentos")

  const armas = loadJsonFiles<Record<string, unknown>>(join(eqDir, "armas"))
  insert("armas", armas.map((a) => ({
    id: a.id, name: a.name,
    category: ARMA_CATEGORY_MAP[a.proficiencia as string] ?? a.proficiencia,
    price: a.preco ?? 0,
    damage: a.dano ?? null,
    damage_type: a.tipo_dano ?? null,
    critical: a.critico ?? null,
    range: a.alcance === "corpo_a_corpo" || !a.alcance ? null : a.alcance,
    weight: a.espacos ?? 0,
    properties: a.habilidades ?? [],
    description: a.description ?? null,
  })))

  const armaduras = loadJsonFiles<Record<string, unknown>>(join(eqDir, "armaduras"))
  insert("armaduras", armaduras.map((a) => ({
    id: a.id, name: a.name, category: a.categoria ?? a.category,
    price: a.preco ?? a.price ?? 0,
    defense_bonus: a.bonus_defesa ?? a.defense_bonus ?? 0,
    armor_penalty: a.penalidade ?? a.armor_penalty ?? 0,
    weight: a.espacos ?? a.weight ?? 0,
    properties: a.habilidades ?? a.properties ?? [],
    description: a.description ?? null,
  })))

  const escudos = loadJsonFiles<Record<string, unknown>>(join(eqDir, "escudos"))
  insert("escudos", escudos.map((e) => ({
    id: e.id, name: e.name,
    price: e.preco ?? e.price ?? 0,
    defense_bonus: e.bonus_defesa ?? e.defense_bonus ?? 0,
    armor_penalty: e.penalidade ?? e.armor_penalty ?? 0,
    weight: e.espacos ?? e.weight ?? 0,
    properties: e.habilidades ?? e.properties ?? [],
    description: e.description ?? null,
  })))

  const itens = loadJsonFiles<Record<string, unknown>>(join(eqDir, "itens"))
  insert("itens", itens.map((i) => ({
    id: i.id, name: i.name, category: i.category ?? null,
    price: i.preco ?? i.price ?? 0,
    weight: i.espacos ?? i.weight ?? 0,
    description: i.description ?? null,
    effects: i.effects ?? {},
  })))
}

function seedItensSuperiores(insert: ReturnType<typeof Database.prototype.transaction>): void {
  const dir = join(JSON_DIR, "itens_superiores")

  const materiais = loadJsonDir<Record<string, unknown>>(join(dir, "materiais_especiais"))
  insert("materiais_especiais", materiais.map((m) => ({
    id: m.id, name: m.name, description: m.description ?? null,
    applicable_to: [], price_modifier: m.precos_adicionais ?? m.price_modifier ?? {},
    effects: m.efeitos ?? m.effects ?? {},
  })))

  const melhorias = loadJsonFiles<Record<string, unknown>>(join(dir, "melhorias"))
  insert("melhorias", melhorias.map((m) => ({
    id: m.id, name: m.name, description: m.description ?? null,
    applicable_to: m.applicable_to ?? [], price: m.price ?? 0,
    effects: m.effects ?? {},
  })))

  const encantamentos = loadJsonFiles<Record<string, unknown>>(join(dir, "encantamentos"))
  insert("encantamentos", encantamentos.map((e) => ({
    id: e.id, name: e.name, description: e.description ?? null,
    categoria: e.categoria ?? "arma",
    escudo_only: e.escudo_only ? 1 : 0,
    conta_como_dois: e.conta_como_dois ? 1 : 0,
    prerequisitos: e.prerequisitos ?? [],
    efeito: e.efeito ?? {},
  })))
}

function seedCondicoes(insert: ReturnType<typeof Database.prototype.transaction>): void {
  const condicoes = loadJsonDir<Record<string, unknown>>(join(JSON_DIR, "condicoes"))
  insert("condicoes", condicoes.map((c) => ({
    id: c.id, name: c.name, description: c.description ?? null,
    effects: c.effects ?? [], condition_type: c.condition_type ?? c.tipo ?? null,
    escalates_to: c.escalates_to ?? null,
  })))
}

function seedPericias(insert: ReturnType<typeof Database.prototype.transaction>): void {
  const pericias = loadJsonDir<Record<string, unknown>>(join(JSON_DIR, "pericias"))
  insert("pericias", pericias.map((p) => ({
    id: p.id, name: p.name, atributo: p.atributo,
    trained_only: p.trained_only ? 1 : 0,
    armor_penalty: p.armor_penalty ? 1 : 0,
    resistance_skill: p.resistance_skill ? 1 : 0,
    description: p.description ?? null, uses: p.uses ?? [],
  })))
}

function seedRegras(insert: ReturnType<typeof Database.prototype.transaction>): void {
  if (!existsDir(join(JSON_DIR, "regras"))) return
  const regras = loadJsonDir<Record<string, unknown>>(join(JSON_DIR, "regras"))
  insert("regras", regras.map((r) => {
    const { id, name, description, ...rest } = r
    return { id, name, description: description ?? null, data: rest }
  }))
}

function existsDir(path: string): boolean {
  try { return statSync(path).isDirectory() } catch { return false }
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  seed()
}
